wgengine/monitor/monitor_windows.go

```diff
@@ -7,273 +7,117 @@ package monitor
-)
-
-const (
-	_STATUS_PENDING = 0x00000103 // 259
-	_STATUS_WAIT_0  = 0
+	errClosed = errors.New("closed")
 )
 
-type unspecifiedMessage struct{}
-
-func (unspecifiedMessage) ignore() bool { return false }
-
-type pollStateChangedMessage struct{}
-
-func (pollStateChangedMessage) ignore() bool { return false }
-
-type messageOrError struct {
-	message
-	error
+type eventMessage struct {
+	eventType string
 }
 
-type winMon struct {
-	ctx             context.Context
-	cancel          context.CancelFunc
-	messagec        chan messageOrError
-	logf            logger.Logf
-	pollTicker      *time.Ticker
-	lastState       *interfaces.State
-	closeHandle     windows.Handle // signaled upon close
-	goroutineDoneCh chan struct{}  // closed when awaitIPAndRouteChanges goroutine has stopped
+func (eventMessage) ignore() bool { return false }
 
-	mu            sync.Mutex
-	lastNetChange time.Time
-	inFastPoll    bool // recent net change event made us go into fast polling mode (to detect proxy changes)
+type winMon struct {
+	logf                  logger.Logf
+	ctx                   context.Context
+	cancel                context.CancelFunc
+	messagec              chan eventMessage
+	addressChangeCallback *winipcfg.UnicastAddressChangeCallback
+	routeChangeCallback   *winipcfg.RouteChangeCallback
 }
 
 func newOSMon(logf logger.Logf, _ *Mon) (osMon, error) {
-	closeHandle, err := windows.CreateEvent(nil, 1 /* manual reset */, 0 /* unsignaled */, nil /* no name */)
+	m := &winMon{
+		logf:     logf,
+		messagec: make(chan eventMessage, 1),
+	}
+
+	var err error
+	m.addressChangeCallback, err = winipcfg.RegisterUnicastAddressChangeCallback(m.unicastAddressChanged)
 	if err != nil {
-		return nil, fmt.Errorf("CreateEvent: %w", err)
+		m.logf("winipcfg.RegisterUnicastAddressChangeCallback error: %v", err)
+		return nil, err
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	m := &winMon{
-		logf:            logf,
-		ctx:             ctx,
-		cancel:          cancel,
-		messagec:        make(chan messageOrError, 1),
-		pollTicker:      time.NewTicker(pollIntervalSlow),
-		closeHandle:     closeHandle,
-		goroutineDoneCh: make(chan struct{}),
+	m.routeChangeCallback, err = winipcfg.RegisterRouteChangeCallback(m.routeChanged)
+	if err != nil {
+		m.addressChangeCallback.Unregister()
+		m.logf("winipcfg.RegisterRouteChangeCallback error: %v", err)
+		return nil, err
 	}
-	go m.awaitIPAndRouteChanges()
+
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
 	return m, nil
 }
 
-func (m *winMon) Close() error {
+func (m *winMon) Close() (ret error) {
 	m.cancel()
-	m.pollTicker.Stop()
-	windows.SetEvent(m.closeHandle) // wakes up any reader blocked in Receive
-
-	// Wait for awaitIPAndRouteChannges to be done, which could
-	// still be using the m.closeHandle.
-	<-m.goroutineDoneCh
 
-	windows.CloseHandle(m.closeHandle)
-	return nil
-}
-
-var errClosed = errors.New("closed")
+	ret = nil
```

Comment:
remove. This isn't C :)  All variables in Go are initialized to their zero value.

Reply:
:)